### PR TITLE
feat: add types to the library

### PR DESCRIPTION
**Pull request by Gozala**
needs:
https://github.com/ipfs/aegir/pull/727
https://github.com/multiformats/js-multihash/pull/113

### CODE REVIEW

File: src/blake.js [hunk 1]
```diff
@@ -21,12 +22,21 @@ const blake2s = {
 // the function as async because it must return a Promise to match the API
 // for other functions that do perform asynchronous work (see sha.browser.js)
 // eslint-disable-next-line
+
+/**
+ * @param {number} size
+ * @param {any} hf
```
Comment (vasco-santos) [inline src/blake.js:28]: nit: perhaps useful to type hf?

File: src/index.js [hunk 2]
```diff
@@ -23,47 +29,49 @@ async function Multihashing (buf, alg, length) {
...
+ * @type {Record<number, Digest>}
  */
+// @ts-ignore - most of those functions aren't typed
 Multihashing.functions = {
   // identity
   0x00: crypto.identity,
   // sha1
   0x11: crypto.sha1,
-  // sha2-256
+  // // sha2-256
```
Comment (vasco-santos) [inline src/index.js:74]:
```suggestion
  // sha2-256
```

File: src/types.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import type { HashName } from 'multihashes'
+
+export interface Digest {
+  (data: Uint8Array): Promise<Uint8Array> | Uint8Array
+}
+
+// This works around the TS bug which emits bad import paths
```
Comment (vasco-santos) [inline src/types.ts:7]: Is this needed after the https://github.com/multiformats/js-multihash/pull/113 followup ? Should we block on that?

**Review by vasco-santos — APPROVED** [review]
Overall this looks good to me! Left a few comments.

This also needs the dependencies linked to be updated once released

File: tsconfig.json [hunk 4]
```diff
@@ -1,8 +1,7 @@
 {
   "extends": "aegir/src/config/tsconfig.aegir.json",
   "compilerOptions": {
-    "outDir": "dist",
-    "noImplicitAny": true
+    "outDir": "dist"
```
Comment (Gozala) [inline tsconfig.json:4]: I think we should keep `"noImplicitAny": true` where we can

```suggestion
    "outDir": "dist"
     "noImplicitAny": true
```

Comment (hugomrdias) [reply tsconfig.json:4]: this is already in the aegir ts config :)

Comment (Gozala) [reply tsconfig.json:4]: I don't see it here https://github.com/ipfs/aegir/blob/master/src/config/tsconfig.aegir.json

Comment (hugomrdias) [reply tsconfig.json:4]: comes default with `strict: true`

Comment (Gozala) [reply tsconfig.json:4]: Nor here https://github.com/ipfs/aegir/blob/feat/docs2.0/src/config/tsconfig.aegir.json which is what this depends on

Comment (Gozala) [reply tsconfig.json:4]: Got it, never mind then.

File: src/blake.js (hunk 1 above)
Comment (hugomrdias) [reply src/blake.js:28]: yes but im going to defer that to another PR

Comment (hugomrdias) [reply src/blake.js:28]: https://github.com/multiformats/js-multihashing-async/issues/96

**Review by achingbrain — APPROVED** [review]

**Pull request merged**